drivers/sensor/MLX90640.zig

```diff
@@ -0,0 +1,882 @@
+        try self.writeThenRead(self.control_register_address, self.frame[0..1]);
+        const val = self.frame[0] >> 7 & 0b111;
+        return @as(u3, @truncate(val));
+    }
+
+    pub fn resolution(self: *Self) !u2 {
+        try self.writeThenRead(self.control_register_address, self.frame[0..1]);
+        const val = self.frame[0] >> 10 & 0b11;
+        return @as(u2, @truncate(val));
+    }
+
+    pub fn temperature(self: *Self, result: []f32) !void {
+        try self.loadFrame();
+
+        const subPage: u16 = self.frame[833] & 0x0001;
+        const vdd = self.getVdd();
+        const ta = self.getTa(vdd);
+        const tr = ta - self.open_air_shift;
+
+        var ta4: f32 = (ta + 273.15);
+        ta4 = ta4 * ta4;
+        ta4 = ta4 * ta4;
+        var tr4: f32 = (tr + 273.15);
+        tr4 = tr4 * tr4;
+        tr4 = tr4 * tr4;
+        const taTr: f32 = tr4 - (tr4 - ta4) / self.emissivity;
+
+        const ktaScale: f32 = @floatFromInt(std.math.pow(u16, 2, self.params.ktaScale));
+        const kvScale: f32 = @floatFromInt(std.math.pow(u16, 2, self.params.kvScale));
+        const alphaScale: f32 = @floatFromInt(std.math.pow(u16, 2, self.params.alphaScale));
+
+        var alphaCorrR: [4]f32 = undefined;
+        alphaCorrR[0] = 1 / (1 + self.params.ksTo[0] * 40);
+        alphaCorrR[1] = 1;
+        var ct: f32 = @floatFromInt(self.params.ct[2]);
+        alphaCorrR[2] = (1 + self.params.ksTo[1] * ct);
+        ct = @floatFromInt(self.params.ct[3] - self.params.ct[2]);
+        alphaCorrR[3] = alphaCorrR[2] * (1 + self.params.ksTo[2] * (ct));
+
+        var gain: f32 = @floatFromInt(self.frame[778]);
+        if (gain > 32767) {
+            gain = gain - 65536;
+        }
+
+        const gee: f32 = @floatFromInt(self.params.gainEE);
+        gain = gee / gain;
+
+        const mode: f32 = @floatFromInt((self.frame[832] & 0x1000) >> 5);
+
+        var irDataCP = [2]f32{
+            @floatFromInt(self.frame[776]),
+            @floatFromInt(self.frame[808]),
+        };
+
+        for (0..2) |i| {
+            if (irDataCP[i] > 32767) {
+                irDataCP[i] = irDataCP[i] - 65536;
+            }
+            irDataCP[i] = irDataCP[i] * gain;
+        }
+
+        var cpo: f32 = @floatFromInt(self.params.cpOffset[0]);
+        irDataCP[0] = irDataCP[0] - cpo * (1 + self.params.cpKta * (ta - 25)) * (1 + self.params.cpKv * (vdd - 3.3));
+
+        const cmee: f32 = @floatFromInt(self.params.calibrationModeEE);
+        cpo = @floatFromInt(self.params.cpOffset[1]);
+        if (mode == cmee) {
+            irDataCP[1] = irDataCP[1] - cpo * (1 + self.params.cpKta * (ta - 25)) * (1 + self.params.cpKv * (vdd - 3.3));
+        } else {
+            irDataCP[1] = irDataCP[1] - (cpo + self.params.ilChessC[0]) * (1 + self.params.cpKta * (ta - 25)) * (1 + self.params.cpKv * (vdd - 3.3));
+        }
+
+        var range: u8 = 0;
+        var pattern: i32 = 0;
+        var pixelNumber: i32 = 0;
+        for (0..768) |i| {
```

Comment:
What is 768? Would like to avoid magic numbers

Reply:
This is the resolution? 24*32? Might want to put these into constants.

Reply:
Yes, this camera has 24x32 pixels.  I do plan on putting them as constants, just trying to think of a name (which is the hard part).

Reply:
WIDTH, HEIGHT, NUM_PIXELS = WIDTH * HEIGHT?